File: Examples/pascal/08_calculator.pas
{
  Simple Calculator
  
  Complete program demonstrating:
  - Menu-driven interface
  - Functions and procedures
  - Case statement
  - Loops for continuous operation
  - Input validation
}

Program Calculator;
Var
  choice: Integer;
  num1, num2, result: Real;
  continueCalc: Char;

{ Function: Add two numbers }
Function Add(a, b: Real): Real;
Begin
  Add := a + b;
End;

{ Function: Subtract two numbers }
Function Subtract(a, b: Real): Real;
Begin
  Subtract := a - b;
End;

{ Function: Multiply two numbers }
Function Multiply(a, b: Real): Real;
Begin
  Multiply := a * b;
End;

{ Function: Divide two numbers }
Function Divide(a, b: Real): Real;
Begin
  Divide := a / b;
End;

{ Procedure: Display menu }
Procedure DisplayMenu;
Begin
  WriteLn('Choose an operation:');
  WriteLn('1. Addition (+)');
  WriteLn('2. Subtraction (-)');
  WriteLn('3. Multiplication (×)');
  WriteLn('4. Division (÷)');
End;

{ Main program }
Begin
  WriteLn('=================================');
  WriteLn('   SIMPLE CALCULATOR PROGRAM');
  WriteLn('=================================');
  WriteLn;
  
  Repeat
    DisplayMenu;
    Write('Enter your choice (1-4): ');
    ReadLn(choice);
    
    If (choice < 1) Or (choice > 4) Then
    Begin
      WriteLn('Invalid choice! Please try again.');
      WriteLn;
      Continue;
    End;
    
    Write('Enter first number: ');
    ReadLn(num1);
    Write('Enter second number: ');
    ReadLn(num2);
    
    Case choice Of
      1: Begin
           result := Add(num1, num2);
           WriteLn;
           WriteLn(num1:0:2, ' + ', num2:0:2, ' = ', result:0:2);
         End;
      2: Begin
           result := Subtract(num1, num2);
           WriteLn;
           WriteLn(num1:0:2, ' - ', num2:0:2, ' = ', result:0:2);
         End;
      3: Begin
           result := Multiply(num1, num2);
           WriteLn;
           WriteLn(num1:0:2, ' × ', num2:0:2, ' = ', result:0:2);
         End;
      4: Begin
           If num2 = 0 Then
             WriteLn('Error: Cannot divide by zero!')
           Else
           Begin
             result := Divide(num1, num2);
             WriteLn;
             WriteLn(num1:0:2, ' ÷ ', num2:0:2, ' = ', result:0:2);
           End;
         End;
    End;
    
    WriteLn;
    Write('Continue? (y/n): ');
    ReadLn(continueCalc);
    WriteLn;
    
  Until (continueCalc <> 'y') And (continueCalc <> 'Y');
  
  WriteLn('Thank you for using the calculator!');
End.

File: Examples/c/08_calculator.c
/*
 * Simple Calculator
 * 
 * Complete program demonstrating:
 * - Menu-driven interface
 * - Functions for operations
 * - Input validation
 * - Switch statement
 * - Loops for continuous operation
 */

#include <stdio.h>

// Function prototypes
float add(float a, float b);
float subtract(float a, float b);
float multiply(float a, float b);
float divide(float a, float b);
void display_menu();

int main() {
    int choice;
    float num1, num2, result;
    char continue_calc;
    
    printf("=================================\n");
    printf("   SIMPLE CALCULATOR PROGRAM\n");
    printf("=================================\n\n");
    
    do {
        display_menu();
        printf("Enter your choice (1-4): ");
        scanf("%d", &choice);
        
        if (choice < 1 || choice > 4) {
            printf("Invalid choice! Please try again.\n\n");
            continue;
        }
        
        printf("Enter first number: ");
        scanf("%f", &num1);
        printf("Enter second number: ");
        scanf("%f", &num2);
        
        switch (choice) {
            case 1:
                result = add(num1, num2);
                printf("\n%.2f + %.2f = %.2f\n", num1, num2, result);
                break;
            case 2:
                result = subtract(num1, num2);
                printf("\n%.2f - %.2f = %.2f\n", num1, num2, result);
                break;
            case 3:
                result = multiply(num1, num2);
                printf("\n%.2f × %.2f = %.2f\n", num1, num2, result);
                break;
            case 4:
                if (num2 == 0) {
                    printf("\nError: Cannot divide by zero!\n");
                } else {
                    result = divide(num1, num2);
                    printf("\n%.2f ÷ %.2f = %.2f\n", num1, num2, result);
                }
                break;
        }
        
        printf("\nContinue? (y/n): ");
        scanf(" %c", &continue_calc);
        printf("\n");
        
    } while (continue_calc == 'y' || continue_calc == 'Y');
    
    printf("Thank you for using the calculator!\n");
    return 0;
}

void display_menu() {
    printf("Choose an operation:\n");
    printf("1. Addition (+)\n");
    printf("2. Subtraction (-)\n");
    printf("3. Multiplication (×)\n");
    printf("4. Division (÷)\n");
}

float add(float a, float b) {
    return a + b;
}

float subtract(float a, float b) {
    return a - b;
}

float multiply(float a, float b) {
    return a * b;
}

float divide(float a, float b) {
    return a / b;
}
